Add support for stream writer and reader

Asutorufa opened the pull request.

Asutorufa committed:
support stream writer and reader

Signed-off-by: Asutorufa <16442314+Asutorufa@users.noreply.github.com>

Luzifer commented on go.sum:
```diff
@@ -0,0 +1,39 @@
+golang.org/x/sync v0.0.0-20190423024810-112230192c58/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
+golang.org/x/sync v0.0.0-20220722155255-886fb9371eb4/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
+golang.org/x/sync v0.1.0/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
+golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
```
Please clean this file (`go mod tidy`)

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	passphrase []byte
+}
+
+func NewReader(passphrase string, r io.Reader, cg CredsGenerator) *DecryptReader {
```
IMHO a better order would be `io.Reader, string, CredsGenerator` as the reader is the main component and passphrase and cg are only "helpers" for reading. [optional]

Luzifer replied:
Also please add documentation comments for all exported functions.

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	buf        []byte
+}
+
+func NewWriter(passphrase string, w io.Writer, cg CredsGenerator) *EncryptWriter {
```
See comment for parameter order in `NewReader`

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	}
+}
+
+func (d *DecryptReader) Init() error {
```
Please unexport (I don't see any reason this should be exposed to the user of this library as it's internal functionality.) - Also lets maybe call it `initMode()` to differentiate from `init()` functions.

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	}
+}
+
+func (e *EncryptWriter) Init() error {
```
Please unexport (I don't see any reason this should be exposed to the user of this library as it's internal functionality.) - Also lets maybe call it `initMode()` to differentiate from `init()` functions.

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+
+	size := (len(b) / aes.BlockSize) * aes.BlockSize
+
+	// log.Println(size, len(b))
```
Remove commented out code

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+
+	if _, err = d.r.Peek(aes.BlockSize); errors.Is(err, io.EOF) {
+		n = n - int(b[n-1])
+	}
```
Please add an explanatory comment to this code explaining what exactly the intention is behind this in order to spare future readers of this code to figure out this has to do with the PKCS7 padding… :D

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	salt := make([]byte, 8) // Generate an 8 byte salt
+	_, err := io.ReadFull(rand.Reader, salt)
+	if err != nil {
+		return err
```
Please wrap errors originating in third-party libraries (this includes the std-lib)

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	originSize := len(b)
+
+	if e.buf != nil {
+		b = append(e.buf, b...)
```
Is it wise to modify the input variable? This might cause trouble to the surrounding code…

IMHO we always should `tmpB := append(e.buf, b...)` as a nil slice is valid for appending and afterwards work with `tmpB`.

Luzifer commented on stream.go:
```diff
@@ -0,0 +1,177 @@
+	return originSize, err
+}
+
+func (e *EncryptWriter) Flush() error {
```
When adding the doc-comment for this please stress this **MUST** be called in order to have all data written.

Also: Would it maybe make sense to implement a `io.ReadCloser` and do this inside the `Close() error` function? I think that might be more commonly used in the Go world…

Luzifer commented on stream_test.go:
```diff
@@ -0,0 +1,60 @@
+	"testing"
+)
+
+func NoError(t *testing.T, err error) {
```
Please unexport. (And maybe call it `requireNoError` - or use `require.NoError` from testify)

Luzifer commented on stream_test.go:
```diff
@@ -0,0 +1,60 @@
+	t.Log(len(data), buf.Len(), bytes.Equal(data, buf.Bytes()))
+
+	_, err = io.Copy(os.Stdout, NewReader(pass, bytes.NewReader(data), BytesToKeyMD5))
+	os.Stdout.Write([]byte("\n"))
```
Please **do not** write to `os.Stdout` but do a programmatic check of the contents. The test should be fully automated and not rely on the test execution to log something to stdout and let the user compare it.

Luzifer commented on stream_test.go:
```diff
@@ -0,0 +1,60 @@
+	data, err := o.EncryptBinaryBytes(pass, plaintext, BytesToKeyMD5)
+	NoError(t, err)
+
+	buf := bytes.NewBuffer(nil)
+	es := NewWriter(pass, buf, BytesToKeyMD5)
+	_, err = es.Write(plaintext)
+	NoError(t, err)
+	NoError(t, es.Flush())
```
Why does the `TestReader` test function utilize the `NewWriter`? It should take encrypted data (`data`) and read from it comparing the read data to the `plaintext`.

Luzifer commented on stream_test.go:
```diff
@@ -0,0 +1,60 @@
+	da, err := o.DecryptBinaryBytes(pass, buf.Bytes(), BytesToKeyMD5)
+	NoError(t, err)
+
+	t.Log(string(da))
```
Please don't just log the data but do a programmatic check whether the result contains the expected data.

Luzifer commented on go.mod:
```diff
@@ -2,4 +2,4 @@ module github.com/Luzifer/go-openssl/v4
 
 go 1.14
```
You are using `io.ReadFull` below which in 1.14 was `ioutil.ReadFull`, please bump to `1.20` as minimum supported version.

Asutorufa commented:
thanks, please review again.

Luzifer commented:
```
stream.go:118:27        unconvert   unnecessary conversion
stream.go:143:3         revive      unhandled-error: Unhandled error in call to function buf.Write
stream.go:14:6          revive      exported: exported type DecryptReader should have comment or be unexported
stream.go:147:2         revive      unhandled-error: Unhandled error in call to function buf.Write
stream.go:163:21        wrapcheck   error returned from interface method should be wrapped: sig: func (io.Writer).Write(p []byte) (n int, err error)
stream.go:181:9         wrapcheck   error returned from interface method should be wrapped: sig: func (io.Writer).Write(p []byte) (n int, err error)
stream.go:21:1          revive      exported: exported function NewReader should have comment or be unexported
stream.go:52:10         wrapcheck   error returned from external package is unwrapped: sig: func crypto/aes.NewCipher(key []byte) (crypto/cipher.Block, error)
stream.go:73:13         wrapcheck   error returned from external package is unwrapped: sig: func (*bufio.Reader).Read(p []byte) (n int, err error)
stream.go:80:3          gocritic    assignOp: replace `n = n - int(b[n-1])` with `n -= int(b[n-1])`
stream.go:86:6          revive      exported: exported type EncryptWriter should have comment or be unexported
stream.go:94:1          revive      exported: exported function NewWriter should have comment or be unexported
```

Asutorufa commented:
Please review again.

Luzifer committed:
Implement pull-request feedback

- Document exported functions
- Sort unexported functions below exported
- Return objects instead of interfaces
- Improve documentation of padding detection

Signed-off-by: Knut Ahlers <knut@ahlers.me>

Luzifer commented:
@Asutorufa I've fixed my remaining remarks, please have a look at the changes
